### ISSUE TITLE
Candidates generation

### ISSUE THREAD

**Issue by Hannibal046** [rooshenas/ebr_mt#4]
Hi, which kind of methods do you use to generate candidates ? 
Beam search or multinomial sampling ?
https://github.com/rooshenas/ebr_mt/blob/cc27fbe575cebda523f1327ec38291c21afbe899/src/fairseq/downlaod_data-iwdeen.sh#L38-L40

https://github.com/rooshenas/ebr_mt/issues/3#issuecomment-1146646195

**Comment by Hannibal046** [comment]
Also, what kind of data format should I use to keep training multilingual bert with MLM ? And how to make sure the MLM training only occurs in the target side of `[cls] src [sep] trg` as suggested in the paper section **4.4**. The data collator in `bert_using_trainerE.py` is `DataCollatorForLanguageModeling` which randomly mask tokens across the whole input.
Thanks for your help !
https://github.com/rooshenas/ebr_mt/blob/cc27fbe575cebda523f1327ec38291c21afbe899/src/fairseq/multibert_finetune/run_bert.sh#L4-L5

**Comment by Hannibal046** [comment]
Should this be removed or kept ? Could you please show some intuition behind this loss term ? Thanks so much
https://github.com/rooshenas/ebr_mt/blob/cc27fbe575cebda523f1327ec38291c21afbe899/src/fairseq/train_ebr.py#L324

**Comment by sumantakcs** [comment]
Hi,

Yes the reported results have been taken after candidate generation through multinomial sampling but i have also checked sampling through diferent methods (i.e. beam search, stochastic beam search, top p,top k etc). I got the best candidates using multinomial sampling but you are free to try different sampling methods. If i remember it correctly rest of the sampling performance kind of work like this beam search<top p<stochastic beam search=<top k (with temperature). Unfortunately, i don't think i have the exact script to generate the candidates through multinomial sampling right now ! Since i lost the track of that particular script while changing the code for other sampling techniques but you can generate it by changing the fairseq decode part or another way i also have used during experiment is openNMT. Although multinomial sampling has a clear support over the transformer distribution hence the performance boost Please note that training and success of this energy-based model's training depend on how much qualitative support you get from the candidates (i.e. in my experience i found most diverse set of samples through stochastic beam search rather than multinomial sampling but the oracle bleu score is low so dropped that.) but it will be interesting to see in case if the stochastic beam search can be exploited to get a comparable oracle bleu score along with diversity how these EBMs work! 

For the loss function you can think of it is a regularization/penalty term so that energy scores get bounded.
